.github/copilot-instructions.md

```diff
@@ -0,0 +1,837 @@
+THANDLE(DATABASE_CONNECTION) database_connection_create(const char* connection_string)
+{
+    THANDLE(DATABASE_CONNECTION) result = THANDLE_MALLOC(DATABASE_CONNECTION)(dispose_function);
+    if (result != NULL)
+    {
+        if (initialize_connection(result, connection_string) != 0)
+        {
+            THANDLE_ASSIGN(DATABASE_CONNECTION)(&result, NULL);  // Clean up on failure
+        }
+    }
+    return result;
+}
+
+// Usage
+THANDLE(DATABASE_CONNECTION) db = database_connection_create("server=localhost");
+if (db != NULL)
+{
+    execute_query(db, "SELECT * FROM users");
+    // Manual cleanup required
+    THANDLE_ASSIGN(DATABASE_CONNECTION)(&db, NULL);
+}
+```
+
+#### Resource Pool with THANDLE
+```c
+typedef struct CONNECTION_POOL_TAG
+{
+    THANDLE(DATABASE_CONNECTION)* connections;
+    size_t pool_size;
+    // ... synchronization primitives
+} CONNECTION_POOL;
+
+THANDLE(DATABASE_CONNECTION) pool_acquire_connection(CONNECTION_POOL* pool)
+{
+    THANDLE(DATABASE_CONNECTION) result = NULL;
+    
+    // Find available connection
+    for (size_t i = 0; i < pool->pool_size; i++)
+    {
+        if (is_connection_available(&pool->connections[i]))
+        {
+            THANDLE_ASSIGN(DATABASE_CONNECTION)(&result, pool->connections[i]);
+            break;
+        }
+    }
+    
+    return result;  // Caller gets shared ownership
+}
+```
+
+### THANDLE Error Handling
+
+#### Safe Initialization Patterns
+```c
+// Always check THANDLE creation results
+THANDLE(MY_OBJECT) safe_create_object(void)
+{
+    THANDLE(MY_OBJECT) result = NULL;
+    
+    result = THANDLE_MALLOC(MY_OBJECT)(dispose_function);
+    if (result == NULL)
+    {
+        LogError("Failed to create MY_OBJECT");
+        goto cleanup;
+    }
+    
+    if (initialize_object_data(result) != 0)
+    {
+        LogError("Failed to initialize object data");
+        THANDLE_ASSIGN(MY_OBJECT)(&result, NULL);  // Clean up
+        goto cleanup;
+    }
+    
+cleanup:
+    return result;
+}
+```
+
+#### Chain of Operations
+```c
+// THANDLE enables safe chaining of operations
+int process_pipeline(THANDLE(INPUT_DATA) input)
+{
+    int result;
+    THANDLE(PROCESSED_DATA) stage1_result = NULL;
+    THANDLE(FINAL_DATA) final_result = NULL;
+    
+    if (input == NULL)
+    {
+        result = -1;
+        goto cleanup;
+    }
+    
+    stage1_result = process_stage_1(input);
+    if (stage1_result == NULL)
+    {
+        result = -1;
+        goto cleanup;
+    }
+    
+    final_result = process_stage_2(stage1_result);
+    if (final_result == NULL)
+    {
+        result = -1;
+        goto cleanup;
+    }
+    
+    result = save_results(final_result);
+    
+cleanup:
+    // All THANDLE variables require manual cleanup if desired
+    return result;
+}
+```
+
+### THANDLE Best Practices
+
+#### ✅ Recommended Patterns
+```c
+// Use THANDLE for all reference-counted objects
+THANDLE(MY_OBJECT) obj = create_object();
+
+// Prefer THANDLE_ASSIGN for sharing
```

Comment:
>Prefer [](http://example.com/codeflow?start=3&length=6)

Not a matter of "preference". It is "the only way" and that's by design.

Reply:
Changed to Use